### PR TITLE
Move "tracks" analytics into a shared calypso-analytics package

### LINKED ISSUE #38799: [Gutenboarding]: add analytics (tracks)
Subtask of EPIC #38701

Along with generating a temporary user account and site to allow users to interact with Gutenboarding, we’ll probably have to set up some way to track it all.

⚠️ We'll need to work out whether we're going to import [lib/analytics](https://github.com/Automattic/wp-calypso/blob/master/client/lib/analytics/index.js) into the project.

From the discussion in p1579140697040600-slack-gutenboarding I think the consensus is that:

- we start with just [tracks](https://github.com/Automattic/wp-calypso/blob/master/client/lib/analytics/index.js#L471) for Gutenboarding
- instead of pulling in the entire `lib/analytics` behemoth, we build a tiny tracks lib in @automattic/analytics package, which will furnish the necessary tracks methods:
   - `recordEvent`, `recordPageView`

### After account creation,  and reassign events triggered by the anonymous ID

For WordPress.com, before users sign up, we set a cookie that assigns to them an anonymous ID.

After sign up, [Calypso calls a method](https://github.com/Automattic/wp-calypso/blob/77318c9b32568459aeb80639f8e01a137b3a5ee1/client/lib/signup/step-actions.js#L481) called [identifyUser](https://github.com/Automattic/wp-calypso/blob/master/client/lib/analytics/index.js#L716) , which records the user’s new ID and username, both of which we send back in the API response after a successful account creation.

Tracks then (somehow) reassigns events triggered by the anonymous ID to the new user ID. 👍

We’d have to determine whether anonymous ID tracking in Gutenboarding before user sign up/sign in is acceptable.

Since it’s working currently on WordPress.com however, my guess is that it is appropriate. A conversion with folks from data might be enough to give us the greet tick.

### Event validation

Double check with the data team that our events and props are okay in light of the product.

## Related links

 p1579140697040600-slack-gutenboarding

Some initial investigation on anonymous IDs: pbxlJb-4t-p2

### PR DESCRIPTION
**Pull request by roo2**
#### Changes proposed in this Pull Request

* Move `tracks` analytics into a shared package `automattic/calypso-analytics`.
* Leave ad-tracking inside of `lib/analytics`, leave GA and refer tracking there for now
* Update existing `lib/analytics` code to use the new `automattic/calypso-analytics` package

#### Testing instructions

- [x] load login page, should make same calls to pixel.wordpress.com as production does, all params present (pageview)
- [x] click login button, should make same calls as production (recordTracksEvent)
- [x] run unit tests (possibly add some where needed)
- [x] Add unit tests for recordTracksEvent
- [x] test Tracks Cookies are initialized
- [x] test GA still initialized (GA is only enabled in prod, not staging, need to enable it in development.json)
- [x] test GA tracking still work for events
- [x] test GA tracking still work on page loads
- [x] test mc bumpstat still works for page views and events (enabled in prod, enable mc_analytics_enabled and restart `npm start`)
- [x] test refer recordPageView
- [x] test when tracks is blocked (blocked `stats.wp.com` with ublock)
- [x] test that current user is being identified with hashed PII 
- [x] test ad tracking, anonymous user id's are still generated correctly
- [x] a test to ensure that subscribers to `analytics.recordEvent` are still notified
- [x] Ensure initializeAnalytics and identifyUser call floodlight (window.gtag in ad-tracking.js) correctly

Fixes #38799

### CODE REVIEW

File: packages/calypso-analytics/package.json
Comment (blowery) [inline packages/calypso-analytics/package.json]: these can all be caret versions (like `^1.1.7`) so we're not over-specifying for consumers

Comment (roo2) [reply packages/calypso-analytics/package.json]: thanks, I've updated that

**Comment by p-jackson** [comment]
rebased

File: packages/calypso-analytics/src/index.js
Comment (p-jackson) [inline packages/calypso-analytics/src/index.js]: It took me a while to figure out how `default as doNotTrack` was working. Do you think it'd be clearer to replace these two lines with a single one:

```js
export { doNotTrack, getCurrentUser, setCurrentUser };
```

File: client/lib/analytics/hotjar.js
Comment (p-jackson) [inline client/lib/analytics/hotjar.js]: I think this line should move to `External dependencies`. Also appears in a few other places.

File: packages/calypso-analytics/README.md
Comment (p-jackson) [inline packages/calypso-analytics/README.md]: Maybe you got throw in a brief word about how this only includes track events. And a link to `client/lib/analytics/docs/tracks.md`?

File: client/reader/search-stream/suggestion-provider.jsx
Comment (p-jackson) [inline client/reader/search-stream/suggestion-provider.jsx]: I don't see where this function is exported as a default ...

File: client/lib/analytics/hotjar.js
Comment (p-jackson) [reply client/lib/analytics/hotjar.js]: I'm just checking that p4TIVU-9lc-p2

File: packages/calypso-analytics/src/index.js
Comment (ramonjd) [inline packages/calypso-analytics/src/index.js]: This should be exported I think?

File: client/reader/search-stream/suggestion-provider.jsx
Comment (ramonjd) [reply client/reader/search-stream/suggestion-provider.jsx]: Probably should be exported from https://github.com/Automattic/wp-calypso/pull/39001/files#diff-653824456d7b9c1440ed064c9ad1abecR65, then imported like so: ?

`import { createRandomId } from '@automattic/calypso-analytics';
`

File: client/lib/analytics/index.js
Comment (jsnajdr) [inline client/lib/analytics/index.js]: The `record-event` event has only one usage: reporting the last tracks event as part of debug info when the JS app crashes.

That could be replaced with a `getLastTracksEvent()` function that's part of the tracks API.

And maybe it could be removed altogether -- I'm not sure if we use that info for anything and I don't see why it would be interesting when debugging JS crashes.

File: packages/calypso-analytics/src/index.js
Comment (jsnajdr) [inline packages/calypso-analytics/src/index.js]: This `src/index.js` file feels very disorganized -- a random heap of functions and constants rather than a clean package API.

Can we move all Tracks code to a dedicated module, e.g., `src/tracks.js` and use `src/index.js` only for re-exporting things?

File: client/lib/analytics/ad-tracking.js
Comment (roo2) [reply client/lib/analytics/ad-tracking.js]: I changed it to `getDoNotTrack` to make it clear that its just reading the value

File: packages/calypso-analytics/src/index.js
Comment (roo2) [reply packages/calypso-analytics/src/index.js]: I made that change and it looks much better :)

File: client/lib/analytics/index.js
Comment (roo2) [reply client/lib/analytics/index.js]: Hmm I'd rather not take this out in this PR, so the PR can stay pure refactoring.

**Comment by ramonjd** [comment]
Was this a wonky rebase? That commit list looks humungous!  :)

**Comment by roo2** [comment]
> Was this a wonky rebase? That commit list looks humungous! :)

Ahh yes, I had the base set wrong 🤦‍♂

File: client/lib/analytics/index.js
Comment (roo2) [reply client/lib/analytics/index.js]: I did investigate it and I believe adding the `getLastTracksEvent` would work but I saw it was possible that other events are added also using the `recordEvent` name so the behavior might change slightly.. It probably wouldn't be an issue to anyone but I don't want to take on the burden of testign it etc.

File: client/lib/analytics/test/index.js
Comment (ramonjd) [inline client/lib/analytics/test/index.js]: No need to change anything here, but if you don't feel like using `Promise.then` you could use `async () => {`, then 

```
await analytics.initialize();
expect( loadScript ).toHaveBeenCalledWith( expect.stringMatching( /\/nostats.js/ ) );
```

File: packages/calypso-analytics/src/index.js
Comment (ramonjd) [inline packages/calypso-analytics/src/index.js]: Looks like `getDoNotTrack` is the default export: https://github.com/Automattic/wp-calypso/pull/39001/files#diff-6dd963982aebbb28e8af6ba227530323R11

File: client/lib/analytics/hotjar.js
Comment (ramonjd) [inline client/lib/analytics/hotjar.js]: Looks like `getDoNotTrack` is the default export: https://github.com/Automattic/wp-calypso/pull/39001/files#diff-6dd963982aebbb28e8af6ba227530323R11

File: client/lib/analytics/ad-tracking.js
Comment (ramonjd) [inline client/lib/analytics/ad-tracking.js]: Check export. Looks like it's the default in the module.

File: packages/calypso-analytics/src/index.js
Comment (ramonjd) [reply packages/calypso-analytics/src/index.js]:
```
export { default as getDoNotTrack } from './utils/do-not-track';
```

Seems to work for me

Comment (roo2) [reply packages/calypso-analytics/src/index.js]: thanks!

**Comment by ramonjd** [comment]
I've smoke tested this patch: created a site, published posts, clicked around Calypso. 

Tracks events are firing without errors. 

Would be good to get a `LGTM` from @Automattic/team-calypso 😇

File: client/reader/search-stream/suggestion-provider.jsx
Comment (jsnajdr) [inline client/reader/search-stream/suggestion-provider.jsx]: I'm not sure if it makes sense to export `createRandomId` from the analytics library. It produces a general-purpose random string. Not specific to analytics in any way.

In this Reader file, it seems to be used for a non-analytics purpose. Or is the railcars code intended for reporting to analytics? @blowery might know more 🙂

File: client/package.json
Comment (jsnajdr) [inline client/package.json]: Let's add the dependency to the root `package.json`, too (and keep it in `client/package.json`). That's needed to ensure reliable NPM installs and lockfiles. See #39218 (invented and merged yesterday) for more details.

**Review by jsnajdr — COMMENTED** [review]
Posted a few easy-to-fix comments. Otherwise, this PR looks good to go.

File: client/reader/search-stream/suggestion-provider.jsx
Comment (roo2) [reply client/reader/search-stream/suggestion-provider.jsx]: ok, I couldn't see a good place to move it to and I think it is a bit small to justify its own package so I've added a new package `calypso-utils` with just createRandomId in it for now, does that sound good to you?

File: client/package.json
Comment (roo2) [reply client/package.json]: ok wow that's good to know I've made that change

File: client/reader/search-stream/suggestion-provider.jsx
Comment (jsnajdr) [reply client/reader/search-stream/suggestion-provider.jsx]: I think that creating a new package is not worth it, especially if the package doesn't have a clear purpose and is just a bag of random utilities.

To be honest, I would just duplicate the code at this moment 🙂 Or find an existing NPM package that generates a random base64 string. `randomstring` or `crypto-random-string` look good.

Comment (roo2) [reply client/reader/search-stream/suggestion-provider.jsx]: ahh true, ok I'll look into one of those packages

Comment (roo2) [reply client/reader/search-stream/suggestion-provider.jsx]: I had a look but I couldn't get the packages working in the browser (they use node crypto), I just went with duplicating the code between those two places it was used

**Comment by ramonjd** [comment]
Think I have a fix for the failing mobile e2e test: https://github.com/Automattic/wp-calypso/pull/39289

**Review by ramonjd — APPROVED** [review]
After you get those quick fixes @jsnajdr mentioned I think this little beauty's good to go

🎉

File: client/reader/search-stream/suggestion-provider.jsx
Comment (jsnajdr) [reply client/reader/search-stream/suggestion-provider.jsx]: > I couldn't get the packages working in the browser 

Oh, that's a bummer that the popular packages are Node.js only 😢 It seems that, after all, there is still a market for a tiny `@automattic/random-string` package 🙂

**Review by jsnajdr — APPROVED** [review]
:shipit:

File: packages/calypso-analytics/README.md
Comment (simison) [inline packages/calypso-analytics/README.md]: Tiiiny nitpick; other packages have their readme headings as `# Package name`, so:

```
# Calypso Analytics
```

File: packages/calypso-analytics/README.md
Comment (simison) [inline packages/calypso-analytics/README.md]: You can grab P2 shorthands for field guide pages, too. I think that would be helpful. Totes not a blocker for merging. :-)

**Comment by simison** [comment]
Any reason for package name to be Calypso specific? `@automattic/analytics` seems like something I'd use in Jetpack, Woo and elsewhere.

**Comment by jsnajdr** [comment]
> Any reason for package name to be Calypso specific?

One way to think about it is that Calypso is the name of the framework the package is a part of 🙂

**Comment by simison** [comment]
>One way to think about it is that Calypso is the name of the framework the package is a part of 🙂

That's fair. :-) (And by no mean I want to block merging this, I don't feel strongly about this).

**Comment by ramonjd** [comment]
In relation to this commit: https://github.com/Automattic/wp-calypso/pull/39001/commits/3f7a7d07971e01564f55acb6a5e5243c145a160b

You can trigger the tests by removing the "Needs review" label and re-adding it 👍

**Pull request merged**
